### PR TITLE
Iterator syntax (among other)

### PR DESCRIPTION
**Pull request by trasher**

### CODE REVIEW

**Comment by trasher** [comment]
To be reviewed per commit I guess ;)

File: source/devapi/database/dbiterator.rst [hunk 1]
```diff
@@ -105,72 +74,89 @@ The fields array can also contain per table sub-array:
...
-                 'FKEY' => ['glpi_computers'=>'id',
-                            'glpi_computerdisks'=>'computer_id']]);
-   $DB->request(['glpi_computers', 'glpi_computerdisks'],
-                ['FKEY' => ['glpi_computers'=>'id',
-                            'glpi_computerdisks'=>'computer_id']]);
+   $DB->request([
+       'FROM' => ['glpi_computers', 'glpi_computerdisks'],
+       'ON' => [
+           'glpi_computers'=>'id',
+           'glpi_computerdisks'=>'computer_id'
```
Comment (cedric-anne) [inline source/devapi/database/dbiterator.rst:89]:
```suggestion
           'glpi_computers' => 'id',
           'glpi_computerdisks' => 'computer_id'
```

File: source/upgradeguides/glpi-10.1.rst [hunk 2]
```diff
@@ -53,3 +53,23 @@ Also, code that ouputs javascript must be adapted to prevent XSS with both HTML
...
+Query builder usage
++++++++++++++++++++
+
+Since it has been implemented, internal query builder (named `DBMysqlIterator`) do accept several syntaxes; that make things complex:
+
+1. conditions (including table name as `FROM` array key) as first (and only) parameter.
+2. table name as first parameter and condition as second parameter,
+3. raw SQL queries,
+
+The most logical one was only the first one, therefore all other has now been deprecated.
```
Comment (cedric-anne) [inline source/upgradeguides/glpi-10.1.rst:66]:
```suggestion
The mostly used and the easier to maintain was the first one. The second one has been deprecated and the thrird one has been prohibited or security reasons.
```

File: source/upgradeguides/glpi-10.1.rst [hunk 3]
```diff
@@ -53,3 +53,23 @@ Also, code that ouputs javascript must be adapted to prevent XSS with both HTML
...
+
+I you were using the second syntax,you will replace as follow:
```
Comment (cedric-anne) [inline source/upgradeguides/glpi-10.1.rst:68]:
```suggestion
I you were using the second syntax, you will need to replace as follows:
```

**Review by cedric-anne — APPROVED** [review]

File: source/devapi/database/dbiterator.rst [hunk 4]
```diff
@@ -5,6 +5,7 @@ GLPI framework provides a simple request generator:
 
 * without having to write SQL
 * without having to quote table and field name
+* without escaping values to prevent SQL injections
```
Comment (cedric-anne) [inline source/devapi/database/dbiterator.rst:8]:
```suggestion
* without having to escape values to prevent SQL injections
```

**Review by cedric-anne — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

File: source/devapi/database/dbiterator.rst [hunk 5]
```diff
@@ -105,72 +74,89 @@ The fields array can also contain per table sub-array:
...
 
    <?php
-   $DB->request(['FROM' => ['glpi_computers', 'glpi_computerdisks'],
-                 'FKEY' => ['glpi_computers'=>'id',
-                            'glpi_computerdisks'=>'computer_id']]);
-   $DB->request(['glpi_computers', 'glpi_computerdisks'],
-                ['FKEY' => ['glpi_computers'=>'id',
-                            'glpi_computerdisks'=>'computer_id']]);
+   $DB->request([
+       'FROM' => ['glpi_computers', 'glpi_computerdisks'],
```
Comment (cconard96) [inline source/devapi/database/dbiterator.rst:86]: Not really related to this PR, but I wonder if implicit joins should be discouraged/deprecated. Not every DBMS supports them, and it is the "old" way to handle joins.

File: source/devapi/database/dbiterator.rst [hunk 6]
```diff
@@ -339,11 +318,25 @@ Using the ``OR``, ``AND``, or ``NOT`` option with an array of criteria:
...
+       ]
+    ]);
    // => SELECT * FROM `glpi_computers` WHERE (`is_deleted` = 0 OR `name` = 'foo')"
 
-   $DB->request('glpi_computers', ['NOT' => ['id' => [1,2,7]]]);
+   $DB->request([
+       'FROM' => 'glpi_computers',
+       'WHERE' => [
+           'NOT' => [
+               'id' => [1,2,7]
```
Comment (cconard96) [inline source/devapi/database/dbiterator.rst:336]:
```suggestion
               'id' => [1, 2, 7]
```

File: source/devapi/database/dbiterator.rst [hunk 7]
```diff
@@ -533,3 +576,15 @@ You can use a QueryExpression object in the FROM statement:
...
    // => SELECT * FROM (SELECT * FROM glpi_computers) as computers
+
+.. warning::
+
+   If you really cannot use any of the above, you still can make raw SQL queries:
+
+   .. code-block:: php
+
+      <?php
+      $DB->doQuery('SHOW COLUMNS FROM `glpi_computers`');
```
Comment (cconard96) [inline source/devapi/database/dbiterator.rst:587]:
```suggestion
      $DB->doQuery('SHOW COLUMNS FROM ' . $DB::quoteName('glpi_computers'));
```

**Review by cconard96 — CHANGES_REQUESTED** [review]

File: source/devapi/database/dbiterator.rst (hunk 5 above)
Comment (cedric-anne) [reply source/devapi/database/dbiterator.rst:86]: It should probably be deprecated. I remember that they are used at some places in GLPI, but probably only on a few cases that can be easilly refactored.

Comment (trasher) [reply source/devapi/database/dbiterator.rst:86]: I'll remove that from documentation

**Review by cconard96 — APPROVED** [review]

**Pull request merged**